### PR TITLE
fixes #9773 - correctly render template URL

### PR DESCRIPTION
**Pull request by stbenjam**
This fixes the URL rendering incorrectly as `https://http://foo.bar.baz` and also ensures the test would catch such a failure in the future.

### CODE REVIEW

File: app/controllers/concerns/foreman/controller/config_templates.rb [hunk 1]
```diff
@@ -13,8 +13,9 @@ def load_vars_from_config_template
   private
 
   def default_template_url(template, hostgroup)
-    url_for(:host => Setting[:unattended_url], :action => :template, :controller => '/unattended',
-            :id => template.name, :hostgroup => hostgroup.name)
+    path = url_for(:only_path => true, :action => :template, :controller => '/unattended',
+                   :id => template.name, :hostgroup => hostgroup.name)
+    Setting[:unattended_url] + path
```
Comment (lzap) [inline app/controllers/concerns/foreman/controller/config_templates.rb]: This is vulnerable to trailing slash, can you fix this and add a test for it?

Comment (stbenjam) [inline app/controllers/concerns/foreman/controller/config_templates.rb]: Sorry what do you mean?

The URL renders correctly:
`ks=http://katello-dev.bitbin.de/unattended/template/Katello%20Kickstart%20Default/With%20Puppet`

Comment (lzap) [inline app/controllers/concerns/foreman/controller/config_templates.rb]: Yeah I have just verified.

File: test/functional/config_templates_controller_test.rb [hunk 2]
```diff
@@ -79,7 +79,7 @@ class ConfigTemplatesControllerTest < ActionController::TestCase
       File.read(File.expand_path(File.dirname(__FILE__) + "/../../app/views/unattended/pxe/PXELinux_default.erb")))
     Setting[:unattended_url] = "http://foreman.unattended.url"
 
-    ProxyAPI::TFTP.any_instance.expects(:create_default).with(has_entry(:menu, regexp_matches(/http:\/\/foreman.unattended.url/))).returns(true)
+    ProxyAPI::TFTP.any_instance.expects(:create_default).with(has_entry(:menu, regexp_matches(/ks=http:\/\/foreman.unattended.url/))).returns(true)
```
Comment (lzap) [inline test/functional/config_templates_controller_test.rb]: For completeness, can we even stronger regexp here? This is how things render in my case

```
append initrd=boot/CentOS-6.5-x86_64-initrd.img ks=http://192.168.122.1/unattended/template/Kickstart%20default/with-root-pw ksdevice=bootif network kssendmac
```

so we can match for

```
/ks=http:\/\/foreman.unattended.url\/template\//
```

**Comment by lzap** [comment]
Ping? @stbenjam

File: app/views/unattended/pxe/PXELinux_default.erb [hunk 3]
```diff
@@ -28,9 +28,9 @@ LABEL <%= "#{profile[:template]} - #{profile[:hostgroup]}" %>
...
+     APPEND initrd=<%= profile[:hostgroup].operatingsystem.initrd(profile[:hostgroup].architecture) %> ks=<%= default_template_url(profile[:template], profile[:hostgroup], @proxy) %> ksdevice=bootif network kssendmac
 <% when 'preseed' -%>
-     APPEND initrd=<%= profile[:hostgroup].operatingsystem.initrd(profile[:hostgroup].architecture) %> interface=auto url=<%= default_template_url(profile[:template], profile[:hostgroup]) %> ramdisk_size=10800 root=/dev/rd/0 rw auto hostname=unassigned-hostname locale=<%= profile[:hostgroup].params['lang'] || 'en_US' %> console-setup/ask_detect=false console-setup/layout=USA console-setup/variant=USA
+     APPEND initrd=<%= profile[:hostgroup].operatingsystem.initrd(profile[:hostgroup].architecture) %> interface=auto url=<%= default_template_url(profile[:template], profile[:hostgroup], @proxy) %> ramdisk_size=10800 root=/dev/rd/0 rw auto hostname=unassigned-hostname locale=<%= profile[:hostgroup].params['lang'] || 'en_US' %> console-setup/ask_detect=false console-setup/layout=USA console-setup/variant=USA
 <% end -%>
 
 <% end -%>
```
Comment (stbenjam) [inline app/views/unattended/pxe/PXELinux_default.erb:36]: This of course needs to go to community-templates.

**Comment by stbenjam** [comment]
@lzap Sorry, it took a while to come back to this.  I realized I wasn't taking into account the Templates proxy, and needed to make sure this was isolated.  It's ready for another look

**Comment by lzap** [comment]
Hmm it looks like the patch got evolved into something else. We are adding proxy support instead of only fixing "http" prefix. Since there are some downstream bugs, do you want to split this into two commits? I will leave the decision up to you. In case you prefer one commit, reword the commit message a bit.

File: app/helpers/smart_proxies_helper.rb [hunk 4]
```diff
@@ -1,13 +1,13 @@
 module SmartProxiesHelper
   def proxy_actions(proxy, authorizer)
-    [if proxy.features.detect{|f| f.name == "Puppet CA"}
+    [if proxy.has_puppet_ca?
```
Comment (lzap) [inline app/helpers/smart_proxies_helper.rb]: Nice. I see at least one more intance that can be replaced?

```
# ack feature | grep include
app/models/subnet.rb:182:    return unless proxy.features.include?(Feature.find_by_name("DHCP"))

Maybe:
app/models/nic/bmc.rb:17:        proxy = subnet.proxies.select { |subnet_proxy| subnet_proxy.features.map(&:name).include?("BMC") }.first
app/models/smart_proxy.rb:87:    if feature_names.include?('Puppet') && feature_names.include?('Puppet CA')
app/models/smart_proxy.rb:89:    elsif feature_names.include?('Puppet')
app/models/smart_proxy.rb:91:    elsif feature_names.include?('Puppet CA')
lib/foreman/renderer.rb:39:            elsif proxy.present? && proxy.try(:features).map(&:name).include?('Templates') && @host.try(:token).present?
```

File: app/models/config_template.rb [hunk 5]
```diff
@@ -104,20 +104,15 @@ def self.build_pxe_default(renderer)
...
-
     return [422, error_msg] unless error_msg.empty?
 
+    @profiles = pxe_default_combos
+
     error_msgs = []
     proxies.each do |proxy|
       begin
+        menu = renderer.render_safe(default_template.template, [:default_template_url], {:profiles => @profiles, :proxy => proxy})
+
```
Comment (lzap) [inline app/models/config_template.rb]: What was your motivation for this change? Is this related?

File: app/views/unattended/pxe/PXELinux_default.erb (hunk 3 above)
Comment (lzap) [inline app/views/unattended/pxe/PXELinux_default.erb:36]: Can you file the change and link it with this PR before I merge please?

File: test/unit/smart_proxy_test.rb [hunk 6]
```diff
@@ -44,6 +44,12 @@ def test_should_honor_legacy_puppet_hostname_false_setting
     assert_equal proxy.to_s, "puppet.example.com"
   end
 
+  test "should repsond to has_<feature>? methods" do
+    proxy = FactoryGirl.create(:template_smart_proxy)
+    assert proxy.has_templates?
+    refute proxy.has_puppet_ca?
+  end
```
Comment (lzap) [inline test/unit/smart_proxy_test.rb]: Nitpick - this should not depend on FactoryGirl defaults...

**Comment by lzap** [comment]
Ok few comments, then I will test and merge! Nice work.

File: app/models/config_template.rb (hunk 5 above)
Comment (stbenjam) [inline app/models/config_template.rb]: Yes, I have to render the template for each proxy, before it was done once.

File: test/unit/smart_proxy_test.rb (hunk 6 above)
Comment (stbenjam) [inline test/unit/smart_proxy_test.rb]: Why not? The FactoryGirl templates proxy should have the templates feature.

File: app/helpers/smart_proxies_helper.rb (hunk 4 above)
Comment (stbenjam) [inline app/helpers/smart_proxies_helper.rb]: Done

**Comment by stbenjam** [comment]
> Hmm it looks like the patch got evolved into something else. We are adding proxy support instead of only fixing "http" prefix. Since there are some downstream bugs, do you want to split this into two commits? I will leave the decision up to you. In case you prefer one commit, reword the commit message a bit.

Yup, sorry, I split it into two commits, they are really separate I guess.  I had capsule isolation on my mind, and it occurred to me as I was fixing this, that this would only ever use the unattended_url

File: app/views/unattended/pxe/PXELinux_default.erb (hunk 3 above)
Comment (stbenjam) [inline app/views/unattended/pxe/PXELinux_default.erb:36]: https://github.com/theforeman/community-templates/pull/174

How do we handle this? They both would need to be merged, and the templates synced to Foreman

File: test/unit/smart_proxy_test.rb (hunk 6 above)
Comment (lzap) [inline test/unit/smart_proxy_test.rb]: The moment we change defaults this test can break. No biggie.

**Comment by lzap** [comment]
All works fine, except the profile. It rendered "path" incorrectly. Or is this expected?

```
append initrd=boot/CentOS-7.0-x86_64-initrd.img ks=http://192.168.122.1:3000/unattended/template/Kickstart%20default/with-root-pw?path= ksdevice=bootif network kssendmac
```

**Comment by lzap** [comment]
Tests failed :-(

**Comment by stbenjam** [comment]
[test]

Looks like just a transient jenkins error, it couldn't check out the repo.

**Comment by stbenjam** [comment]
> All works fine, except the profile. It rendered "path" incorrectly. Or is this expected?

@lzap Fixed the "?path=" thingy, you shouldn't see that appended anymore

**Comment by lzap** [comment]
It works, once the test is green ready to merge: http://ci.theforeman.org/job/test_develop_pr_core/database=mysql,ruby=2.0.0/lastCompletedBuild/testReport/(root)/UnattendedControllerTest/test_0035_hosts_with_a_template_proxy_with_no_templateServer_should_use_the_proxy_name/

**Comment by stbenjam** [comment]
@lzap I moved 10286 to https://github.com/theforeman/foreman/pull/2327, and this is ONLY addressing the URL issue.  I can't figure out why ONLY mysql/2.0.0 is failing right now

**Comment by lzap** [comment]
Merged as f7174439285708c3010605230fec16797f3a0763, thank you!
